core/src/commands/publish_datapoint.rs

```diff
@@ -0,0 +1,381 @@
+    };
+    tx_builder.set_context_extension(in_oracle_box.get_box().box_id(), ctx_ext);
+    let tx = tx_builder.build()?;
+    Ok(PublishDataPointAction { tx })
+}
+
+fn compute_new_datapoint(datapoint: i64, old_datapoint: i64) -> i64 {
+    // Difference calc
+    let difference = datapoint as f64 / old_datapoint as f64;
+
+    // If the new datapoint is twice as high, post the new datapoint
+    #[allow(clippy::if_same_then_else)]
+    if difference > 2.00 {
+        datapoint
+    }
+    // If the new datapoint is half, post the new datapoint
+    else if difference < 0.50 {
+        datapoint
+    }
+    // TODO: remove 0.5% cap, kushti asked on TG:
+    // >Lets run 2.0 with no delay in data update in the default data provider
+    // >No, data provider currently cap oracle price change at 0.5 percent per epoch
+    //
+    // If the new datapoint is 0.49% to 50% lower, post 0.49% lower than old
+    else if difference < 0.9951 {
+        (old_datapoint as f64 * 0.9951) as i64
+    }
+    // If the new datapoint is 0.49% to 100% higher, post 0.49% higher than old
+    else if difference > 1.0049 {
+        (old_datapoint as f64 * 1.0049) as i64
+    }
+    // Else if the difference is within 0.49% either way, post the new datapoint
+    else {
+        datapoint
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::convert::TryFrom;
+    use std::convert::TryInto;
+
+    use ergo_lib::chain::ergo_state_context::ErgoStateContext;
+    use ergo_lib::chain::transaction::unsigned::UnsignedTransaction;
+    use ergo_lib::chain::transaction::TxId;
+    use ergo_lib::chain::transaction::TxIoVec;
+    use ergo_lib::ergotree_interpreter::sigma_protocol::private_input::DlogProverInput;
+    use ergo_lib::ergotree_ir::chain::address::AddressEncoder;
+    use ergo_lib::ergotree_ir::chain::ergo_box::box_value::BoxValue;
+    use ergo_lib::ergotree_ir::chain::ergo_box::ErgoBox;
+    use ergo_lib::ergotree_ir::chain::ergo_box::NonMandatoryRegisterId;
+    use ergo_lib::ergotree_ir::chain::ergo_box::NonMandatoryRegisters;
+    use ergo_lib::ergotree_ir::chain::token::Token;
+    use ergo_lib::ergotree_ir::chain::token::TokenId;
+    use ergo_lib::ergotree_ir::ergo_tree::ErgoTree;
+    use ergo_lib::ergotree_ir::mir::constant::Constant;
+    use ergo_lib::ergotree_ir::mir::expr::Expr;
+    use ergo_lib::ergotree_ir::sigma_protocol::dlog_group::EcPoint;
+    use ergo_lib::ergotree_ir::sigma_protocol::sigma_boolean::ProveDlog;
+    use ergo_lib::wallet::signing::TransactionContext;
+    use ergo_lib::wallet::Wallet;
+    use ergo_node_interface::node_interface::NodeError;
+    use sigma_test_util::force_any_val;
+
+    use crate::box_kind::OracleBoxWrapper;
+    use crate::box_kind::PoolBoxWrapper;
+    use crate::contracts::oracle::OracleContract;
+    use crate::contracts::pool::PoolContract;
+    use crate::contracts::refresh::RefreshContract;
+    use crate::oracle_state::StageError;
+
+    use super::*;
+
+    #[derive(Clone)]
+    struct PoolBoxMock {
```

Comment:
Seems like a lot is overlapping with mocks and box creation helpers in refresh action tests. Could you extract common code into a separate module and re-use it in both action tests? I suspect we're going to re-use them in the upcoming actions as well.

Reply:
Sure, will do.